dd-sdk-android/src/main/java/com/datadog/opentracing/propagation/B3MHttpCodec.java

```diff
@@ -0,0 +1,143 @@
+import java.util.Map;
+
+/**
+ * A codec designed for HTTP transport via headers using B3M headers
+ *
+ * <p>TODO: there is fair amount of code duplication between DatadogHttpCodec and this class,
+ * especially in part where TagContext is handled. We may want to refactor that and avoid special
+ * handling of TagContext in other places (i.e. CompoundExtractor).
+ */
+class B3MHttpCodec {
+
+  private static final String TRACE_ID_KEY = "X-B3-TraceId";
+  private static final String SPAN_ID_KEY = "X-B3-SpanId";
+  private static final String SAMPLING_PRIORITY_KEY = "X-B3-Sampled";
+  private static final String SAMPLING_PRIORITY_ACCEPT = String.valueOf(1);
+  private static final String SAMPLING_PRIORITY_DROP = String.valueOf(0);
+  private static final int HEX_RADIX = 16;
+
+  private B3MHttpCodec() {
+    // This class should not be created. This also makes code coverage checks happy.
+  }
+
+  public static class Injector implements HttpCodec.Injector {
+
+    @Override
+    public void inject(final DDSpanContext context, final TextMapInject carrier) {
+      try {
+        carrier.put(TRACE_ID_KEY, context.getTraceId().toString(HEX_RADIX).toLowerCase(Locale.US));
+        carrier.put(SPAN_ID_KEY, context.getSpanId().toString(HEX_RADIX).toLowerCase(Locale.US));
+
+        if (context.lockSamplingPriority()) {
+          carrier.put(
+              SAMPLING_PRIORITY_KEY, convertSamplingPriority(context.getSamplingPriority()));
+        }
+      } catch (final NumberFormatException e) {
+      }
+    }
+
+    private String convertSamplingPriority(final int samplingPriority) {
+      return samplingPriority > 0 ? SAMPLING_PRIORITY_ACCEPT : SAMPLING_PRIORITY_DROP;
+    }
+  }
+
+  public static class Extractor implements HttpCodec.Extractor {
+
+    private final Map<String, String> taggedHeaders;
+
+    public Extractor(final Map<String, String> taggedHeaders) {
+      this.taggedHeaders = new HashMap<>();
+      for (final Map.Entry<String, String> mapping : taggedHeaders.entrySet()) {
+        this.taggedHeaders.put(mapping.getKey().trim().toLowerCase(Locale.US), mapping.getValue());
+      }
+    }
+
+    @Override
+    public SpanContext extract(final TextMapExtract carrier) {
+      try {
+        Map<String, String> tags = Collections.emptyMap();
+        BigInteger traceId = BigInteger.ZERO;
+        BigInteger spanId = BigInteger.ZERO;
+        int samplingPriority = PrioritySampling.UNSET;
+
+        for (final Map.Entry<String, String> entry : carrier) {
+          final String key = entry.getKey().toLowerCase(Locale.US);
+          final String value = entry.getValue();
+
+          if (value == null) {
+            continue;
+          }
+
+          if (TRACE_ID_KEY.equalsIgnoreCase(key)) {
+            final String trimmedValue;
+            final int length = value.length();
+            if (length > 32) {
+              traceId = BigInteger.ZERO;
+              continue;
+            } else if (length > 16) {
```

Comment:
This confuses me a little bit --

There'a an edge case where a 16 character hex digit can be larger than the supported 63-bits. This may not be an issue if we're always generating the ID.

But if we're potentially extracting these IDs from user set headers (which I'm told we shouldn't be) then we likely need to handle the edge case (though it's possible `validateUIng64BitsID` does that?)

Reply:
This is original code from the Java tracing lib.
`validateUInt64BitsID` only accepts IDs between `0` and `2^64-1`. It throws an exception otherwise.

In Android we do support extracting the IDs from the Request. It's been like this for quite some time.